solutions/cses-1160.mdx

```diff
@@ -2,62 +2,30 @@
 id: cses-1160
 source: CSES
 title: Planets Queries II
-author: Michael Cao, Benjamin Qi
+author: Dustin Miao, Michael Cao, Benjamin Qi
 ---
 
-In this problem, we are given a directed graph and asked $q$ queries for the
-minimum distance between two vertices $u$ and $v$.
+In this problem, we are given a functional directed graph and asked $q$ queries for the minimum distance between two vertices $u$ and $v$, if it exists. 
 
-# Main Idea
+## Solution
 
-The graph is a functional graph, so break it into tree / cycle cases and solve
-each one independently.
+We will decompose the graph into several cases. Note that because each node has one outgoing edge, there must be exactly one cycle in every connected component. 
 
-# Structure of the Graph
+Compress this cycle into a single node. Each connected component is now a tree, rooted at the cycle node. Consider a query for a path from $u$ to $v$. We will assume that $u$ and $v$ are in the same component (otherwise, clearly no path exits and the answer is $-1$). There are four main cases to consider:
 
-While the graph given in the statement can be modeled as a directed graph, we
-can be more specific. Since each vertex has one outgoing edge, the resulting
-graph is a [Functional Graph](/silver/func-graphs).
+1. If $u$ and $v$ are both located within the cycle, then there always a path between them. "Root" the cycle at a node within the cycle, and for each subsequent node, record the distance between that node at the cycle root. We can now process this kind of query in $\mathcal{O}(1)$ by subtracting the cycle distance information accordingly. 
 
-An important property of a functional graph is that it's structure is composed
-of many trees directed towards the root, and a single cycle composed of the
-roots. So, we can break the graph into two cases.
+2. If $u$ is in the cycle and $v$ is not in the cycle (ie in the tree), then there is no path between them. Nodes in the cycle can only reach other nodes in cycle via the edges. 
 
-# Tree Case
+3. If $u$ is in the tree and $v$ is in the cycle, then there is an answer. If we traverse the edges from $u$ until we reach a node on the cycle, then we can reduce the problem to the first case. By itself, this method takes $\mathcal{O}(n)$ but we can speed it up to $\mathcal{O}(\log n)$ using techniques discussed later. 
 
-The first case is if $u$ and $v$ are on the same tree. To check this, we can
-store the root of the tree that each vertex is on in functional graph, and check
-if the roots for $u$ and $v$ are equal.
+4. If $u$ and $v$ are both on the tree, then it suffices to check if $v$ is an ancestor of $u$. Note that each node can only travel to edges that point towards the root. Once again, in the worst case naive solution this may take up to $\mathcal{O}(n)$ but we will speed it up to $\mathcal{O}(\log n)$
```

Comment:
```suggestion
4. If $u$ and $v$ are both on the tree, then it suffices to check if $v$ is an ancestor of $u$. Note that each node can only travel to edges that point towards the root. Once again, in the worst-case naive solution this may take up to $\mathcal{O}(n)$ but it can be sped up to $\mathcal{O}(\log n)$
```